src/main/java/com/gotocompany/depot/maxcompute/schema/partition/PartitioningStrategyFactory.java

```diff
@@ -0,0 +1,55 @@
+package com.gotocompany.depot.maxcompute.schema.partition;
+
+import com.aliyun.odps.type.TypeInfo;
+import com.aliyun.odps.type.TypeInfoFactory;
+import com.google.protobuf.Descriptors;
+import com.gotocompany.depot.config.MaxComputeSinkConfig;
+import com.gotocompany.depot.maxcompute.converter.ConverterOrchestrator;
+import lombok.RequiredArgsConstructor;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@RequiredArgsConstructor
+public class PartitioningStrategyFactory {
+
+    private final ConverterOrchestrator converterOrchestrator;
+    private final MaxComputeSinkConfig maxComputeSinkConfig;
+    private static final Set<TypeInfo> ALLOWED_PARTITION_KEY_TYPE_INFO;
+
+    static {
+        ALLOWED_PARTITION_KEY_TYPE_INFO = new HashSet<>();
+        ALLOWED_PARTITION_KEY_TYPE_INFO.add(TypeInfoFactory.TIMESTAMP_NTZ);
+        ALLOWED_PARTITION_KEY_TYPE_INFO.add(TypeInfoFactory.STRING);
+        ALLOWED_PARTITION_KEY_TYPE_INFO.add(TypeInfoFactory.TINYINT);
+        ALLOWED_PARTITION_KEY_TYPE_INFO.add(TypeInfoFactory.SMALLINT);
+        ALLOWED_PARTITION_KEY_TYPE_INFO.add(TypeInfoFactory.INT);
+        ALLOWED_PARTITION_KEY_TYPE_INFO.add(TypeInfoFactory.BIGINT);
+    }
+
+    public PartitioningStrategy createPartitioningStrategy(Descriptors.Descriptor descriptor) {
+        if (!maxComputeSinkConfig.isTablePartitioningEnabled()) {
+            return null;
+        }
+        String partitionKey = maxComputeSinkConfig.getTablePartitionKey();
+        Descriptors.FieldDescriptor fieldDescriptor = descriptor
+                .findFieldByName(partitionKey);
+        if (fieldDescriptor == null) {
+            throw new IllegalArgumentException("Partition key not found in the descriptor: " + partitionKey);
+        }
+        TypeInfo partitionKeyTypeInfo = converterOrchestrator.convert(fieldDescriptor);
+        checkPartitionTypePrecondition(partitionKeyTypeInfo);
+        if (TypeInfoFactory.TIMESTAMP_NTZ.equals(partitionKeyTypeInfo)) {
+            return new TimestampPartitioningStrategy(maxComputeSinkConfig);
+        } else {
+            return new DefaultPartitioningStrategy(partitionKeyTypeInfo, maxComputeSinkConfig);
+        }
+    }
```

Comment:
The factory uses if-else logic for strategy selection. We should use a Strategy registry pattern where strategies can register themselves, making it more extensible and maintainable.

Reply:
Could you please give me the example of Strategy Registry Pattern? Isn't it just offloading the selection process to another class? 

I see your point about potential extensibility, and it could make sense if the number of strategies grows significantly. But, introducing another registry class seems overkill since we only have 2 strategy at the moment. How about refactoring this only when it's really needed?